### DISCUSSION TITLE
Exercise 4.2 - Trouble getting transforms to work

### DISCUSSION THREAD

**Discussion by memetzgz** [Lightning-AI/dl-fundamentals#60] [Exercises] [answered]
I'm not sure if we were supposed to apply the same transformations to the images in this exercise that we did to the MNIST images, but I did attempt to do so. I got a type error, I think when it was trying to resize the images (based on the trace), saying: TypeError: Tensor is not a torch image.  When I tried to fix that by using Image.fromarray (and some other attempts to turn it into a torch image), I got an error saying it couldn't determine the dtype of the array.  I thought since the Fashion MNIST images were similar to the classic MNIST ones that it would be easy to apply the same transforms, but apparently I was wrong LOL. Any clues you can give me on where I went wrong?  THanks!

**Comment by rasbt** [comment, marked as answer]
Interesting! You mean the following data transforms?

```python
    data_transforms = {
        "train": transforms.Compose(
            [
                transforms.Resize(32),
                transforms.RandomCrop((28, 28)),
                transforms.ToTensor(),
                # normalize images to [-1, 1] range
                transforms.Normalize((0.5,), (0.5,)),
            ]
        ),
        "test": transforms.Compose(
            [
                transforms.Resize(32),
                transforms.CenterCrop((28, 28)),
                transforms.ToTensor(),
                # normalize images to [-1, 1] range
                transforms.Normalize((0.5,), (0.5,)),
            ]
        ),
    }

    train_dataset = MyDataset(
        csv_path="mnist-pngs/new_train.csv",
        img_dir="mnist-pngs/",
        transform=data_transforms["train"],
    )

    train_loader = DataLoader(
        dataset=train_dataset,
        batch_size=32,
        shuffle=True,  # want to shuffle the dataset
        num_workers=2,  # number processes/CPUs to use
    )
```

Maybe try adding a `transforms.ToPILImage(),` before the `Resize`. E.g,


```python
            [
                transforms.ToPILImage(),
                transforms.Resize(32),
                transforms.CenterCrop((28, 28)),
                transforms.ToTensor(),
                # normalize images to [-1, 1] range
                transforms.Normalize((0.5,), (0.5,)),
            ]
```

**Reply by memetzgz** [reply]
Thank you for the tip!  I appreciate the reply.

For some reason Pytorch didn't like the fact that the tensor was 1 dimensional when I tried what you suggested above. BUT that sent me in the right direction! What eventually ended up working was to add the transform to the MyDataset class, as follows:

        if self.transform is not None:
            img = transforms.ToPILImage()(img.reshape(1,28,28))
            img = self.transform(img)

As it turns out, though, the transforms made the model perform worse rather than better 😭 

Clearly the resize and the cropping weren't necessary given the 28x28 size of the images. Just for fun, I tried again with just the normalization transform, that gave about a 1 percentage point boost to accuracy for train, validation, and test sets. So not stellar, but I guess every percent counts.

**Reply by rasbt** [reply]
Glad to hear that you got it to work! Btw I'd say it also all depends on the model and dataset combination. Sometimes too much data augmentation (resizing and random cropping) can hurt if it is too much.
